examples/apps/windows/ServerControl.xaml.cpp

```diff
@@ -0,0 +1,134 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "pch.h"
+#include <utility>
+#include <algorithm>
+#include "ServerControl.xaml.h"
+#include "Factory.h"
+
+using namespace Thread;
+
+using namespace Platform;
+using namespace Windows::Foundation;
+using namespace Windows::Foundation::Collections;
+using namespace Windows::UI;
+using namespace Windows::UI::Xaml;
+using namespace Windows::UI::Xaml::Controls;
+using namespace Windows::UI::Xaml::Controls::Primitives;
+using namespace Windows::UI::Xaml::Data;
+using namespace Windows::UI::Xaml::Input;
+using namespace Windows::UI::Xaml::Media;
+using namespace Windows::UI::Xaml::Navigation;
+
+// The User Control item template is documented at http://go.microsoft.com/fwlink/?LinkId=234236
+
+ServerControl::ServerControl()
+{
+    InitializeComponent();
+
+    ServerPort->Text = DEF_PORT.ToString();
+}
+
+void
+ServerControl::Init(
+    IAsyncThreadNotify^  notify,
+    IMainPageUIElements^ mainPageUIElements)
+{
+    _notify = std::move(notify);
+    _mainPageUIElements = std::move(mainPageUIElements);
+}
+
+void
+ServerControl::ProtocolChanged(
+    Protocol protocol)
+{
+    _protocol = protocol;
+}
+
+void
+ServerControl::Listen_Click(
+    Object^          sender,
+    RoutedEventArgs^ e)
+{
+    try
+    {
+        auto listenerArgs = ref new ListenerArgs();
+
+        listenerArgs->ServerName = ServerName->Text;
+
+        auto serverIP = ServerIP->Text;
+        if (serverIP->IsEmpty())
+        {
+            Exception::CreateException(E_INVALIDARG, "No Server IP");
+        }
+
+        // check valid chars of Ipv6 Address
+        if (std::all_of(serverIP->Data(), serverIP->Data() + serverIP->Length(),
```

Comment:
The lambda is just as a input argument filter to the DatagramSocket::ConnectAsync, StreamSocket::ConnectAsync, DatagramSocket::BindEndpointAsync, or StreamSocketListener::BindEndpointAsync APIs.

These APIs can validate all the input parameters themselves but they are supporting both Ipv6 and Ipv4. I don't want Ipv4 support in this case because it is Ipv6 only in Thread. That is why I want to filter something out.

I don't want it to be a public static method named like ContainOnlyValidIpv6Chars because the name is misleading. This method name is too generic. This function is intended to be just a filter function for input parameter(s) of those specific APIs only.

Another little benefit
A lambda could be optimized easily by the compiler. (Tested under VC++ 2015 / Release / x64)

for lambda with statement -
if ((c >= '0' && c <= '9') || (c == ':') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F'))

assembly
```
00007FF69C5F847B  mov         ecx,ebx  00007FF69C5F847D  lea         r8,[rdi+rbx*2]  
00007FF69C5F8481  cmp         rax,r8  
00007FF69C5F8484  je          Thread::ClientControl::Connect_Click+1FCh (07FF69C5F84ACh)  
00007FF69C5F8486  mov         r9,7E0000007E07FFh  
00007FF69C5F8490  movzx       ecx,word ptr [rdx]  
00007FF69C5F8493  sub         cx,30h  
00007FF69C5F8497  cmp         cx,36h  
00007FF69C5F849B  ja          Thread::ClientControl::Connect_Click+251h (07FF69C5F8501h)  
00007FF69C5F849D  bt          r9,rcx  
00007FF69C5F84A1  jae         Thread::ClientControl::Connect_Click+251h (07FF69C5F8501h)  
00007FF69C5F84A3  add         rdx,2  
00007FF69C5F84A7  cmp         rdx,r8  
00007FF69C5F84AA  jne         Thread::ClientControl::Connect_Click+1E0h (07FF69C5F8490h)
```

for static inline method -
static inline bool ContainOnlyValidIpv6Chars(wchar_t c)
{
    if ((c >= '0' && c <= '9') || (c == ':') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F'))

assembly
```
00007FF69C5F8525  lea         rax,[Thread::ClientControl::ContainOnlyValidIpv6Chars (07FF69C5F79A0h)]  
00007FF69C5F852C  lea         rdx,[rdi+rsi*2]  
00007FF69C5F8530  cmp         rbx,rdx  
00007FF69C5F8533  je          Thread::ClientControl::Connect_Click+294h (07FF69C5F8544h)  
00007FF69C5F8535  movzx       ecx,word ptr [rbx]  
00007FF69C5F8538  call        rax  
00007FF69C5F853A  test        al,al  
00007FF69C5F853C  je          Thread::ClientControl::Connect_Click+2E9h (07FF69C5F8599h)  
00007FF69C5F853E  add         rbx,2  
00007FF69C5F8542  jmp         Thread::ClientControl::Connect_Click+275h (07FF69C5F8525h)  
```

If you prefer to have a help function, I might need to put it into a private section with friend classes to access it because it should not be a method for general purpose.
Do you have a suggested name for it?